core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -769,6 +799,70 @@ private SubPlan optimizePlan(SubPlan plan)
         {
             // Re-optimize plan here based on available runtime statistics.
             // Fragments changed due to re-optimization as well as their downstream stages are expected to be assigned new fragment ids.
+            plan = updateStagesPartitioning(plan);
+            return plan;
+        }
+
+        private SubPlan updateStagesPartitioning(SubPlan plan)
+        {
+            if (!runtimeAdaptivePartitioningEnabled || runtimeAdaptivePartitioningApplied) {
+                return plan;
+            }
+
+            for (SubPlan subPlan : planInTopologicalOrder) {
+                PlanFragment fragment = subPlan.getFragment();
+                if (!hasInputPartitioning(fragment)) {
+                    // no input hash partitioning present
+                    continue;
+                }
+
+                StageId stageId = getStageId(fragment.getId());
+                if (stageExecutions.containsKey(stageId)) {
+                    // already started
+                    continue;
+                }
+
+                IsReadyForExecutionResult isReadyForExecutionResult = isReadyForExecution(subPlan);
+                // caching is not only needed to avoid duplicate calls, but also to avoid the case
+                // that a subPlan that is not ready now but becomes ready when updateStageExecutions
+                isReadyForExecutionCache.put(subPlan, isReadyForExecutionResult);
+                if (!isReadyForExecutionResult.isReadyForExecution()) {
+                    // not ready for execution
+                    continue;
+                }
+
+                // calculate (estimated) input data size to determine if we want to change number of partitions at runtime
+                long totalBytes = 0;
+                for (RemoteSourceNode remoteSourceNode : fragment.getRemoteSourceNodes()) {
+                    if (remoteSourceNode.getExchangeType() != REPLICATE) {
+                        for (PlanFragmentId sourceFragmentId : remoteSourceNode.getSourceFragmentIds()) {
+                            StageId sourceStageId = getStageId(sourceFragmentId);
+                            OutputDataSizeEstimate outputDataSizeEstimate = isReadyForExecutionResult.getSourceOutputSizeEstimates().get(sourceStageId);
+                            verify(outputDataSizeEstimate != null, "outputDataSizeEstimate not found for source stage %s", sourceStageId);
+                            totalBytes += outputDataSizeEstimate.getTotalSizeInBytes();
+                        }
+                    }
+                }
```

Comment:
I have been thinking whether we should use total bytes. Should we maybe use the largest outputDataSize of all sources and go with a more conservative threshold?

Reply:
Discussed with @arhimondr offline, we should estimate memory usage based on query shapes.

Reply:
Please file an issue and add a todo comment linking to the issue.